Add GitHub social provider.

lucyhe opened the pull request:
Tested manually.

Notes:
- GitHub sent friend requests/received friend requests will always appear in the "Friends who have uProxy" category
- This will still allow the Gmail and Facebook to work if the extension is version 0.8.23 and the app is 0.8.24. However, if we were to try and sign in with GitHub on a 0.8.23 extension, the login will fail.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy/1916)
<!-- Reviewable:end -->

lucyhe commented:
![image](https://cloud.githubusercontent.com/assets/8723127/10086681/bc565278-62dd-11e5-8bf8-99a221209cb1.png)

![image](https://cloud.githubusercontent.com/assets/8723127/10086685/c8ad18ea-62dd-11e5-828c-f85502f620ba.png)

![image](https://cloud.githubusercontent.com/assets/8723127/10086691/d21c7fd8-62dd-11e5-87c7-dd51009af2e4.png)

lucyhe commented:
Multiple social networks:

![image](https://cloud.githubusercontent.com/assets/8723127/10086737/09e20c3a-62de-11e5-9a15-81e786b6c461.png)

lucyhe commented:
Inviting friends:

![image](https://cloud.githubusercontent.com/assets/8723127/10086747/17c73974-62de-11e5-9182-855b9483fc33.png)

If your GitHub friend has not logged in to uProxy before:

![image](https://cloud.githubusercontent.com/assets/8723127/10086760/2ce8087e-62de-11e5-949e-4d6537eb494c.png)

Successful send:

![image](https://cloud.githubusercontent.com/assets/8723127/10086766/35aaded2-62de-11e5-99eb-32ba0ad1e8f3.png)

dborkan commented:
A few questions:
* In the first image lucyhe has a photo, but the 2nd image is one of those default pixel things - was this expected?
* Travis + Shippable are currently failing, maybe some merge error?

---

Review status: 0 of 29 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

<sup>**[package.json, line 34 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1916#-K-42SU0vo8F7mwYPxBw)** ([raw file](https://github.com/uproxy/uproxy/blob/3e3606ef0ada52d6d332354041007772afba6964/package.json#L34)):</sup>
Any reason we are including the package this way instead of our usual way with npm modules?

---

<sup>**[src/generic_core/remote-user.ts, line 479 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1916#-K-44de2o02LMPo_Uklm)** ([raw file](https://github.com/uproxy/uproxy/blob/3e3606ef0ada52d6d332354041007772afba6964/src/generic_core/remote-user.ts#L479)):</sup>
state.status might be undefined from old users in storage right?   I think we might have a race condition here when we get the first onUserProfile events for a user that was already in storage:
- social.ts gets the onUserProfile, it calls ```this.getOrAddUser_(userId).update(profile)``` where getOrAddUser_ calls the RemoteUser constructor, and returns the new RemoteUser, but doesn't wait for it to be loaded from storage.  Then update(profile) is called, which will default the status to 'FRIEND'
- then if restoreState is called (after storage load async finishes), this line will get hit, and it will reset the status back to undefined (it may also undo other changes done in .update)

If I'm correct and this race condition has been around for a while, a quick workaround for this pull request could just be to default status to 'FRIEND' here as well as in .update, then we could deal with the race condition in a separate issue if it exists

---

<sup>**[src/generic_ui/polymer/accept-user-invite.ts, line 15 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1916#-K-47daXPVjVEM91eVJo)** ([raw file](https://github.com/uproxy/uproxy/blob/3e3606ef0ada52d6d332354041007772afba6964/src/generic_ui/polymer/accept-user-invite.ts#L15)):</sup>
Is this supposed to be the userId of the person who sent you the invite, or the userId of the person accepting the invite?  Also this is not going to be available for networks other than GitHub, will that be a problem

I was expecting that stuff in networkData wouldn't need to be used by uProxy at all and would only be for social providers.  Since this is needed by uProxy, can it be moved outside of networkData and treated more like networkName, userName, etc in token?

---

<sup>**[src/generic_ui/polymer/accept-user-invite.ts, line 22 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1916#-K-47bC6DqXK17zneMBa)** ([raw file](https://github.com/uproxy/uproxy/blob/3e3606ef0ada52d6d332354041007772afba6964/src/generic_ui/polymer/accept-user-invite.ts#L22)):</sup>
Can you explain what the 2 userIds are? 

---

<sup>**[src/generic_ui/polymer/contact.ts, line 118 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1916#-K-48bmOjpU0xR4f1eKZ)** ([raw file](https://github.com/uproxy/uproxy/blob/3e3606ef0ada52d6d332354041007772afba6964/src/generic_ui/polymer/contact.ts#L118)):</sup>
how is mode available here?  I'd expect to see something like "this.mode = model.globalSettings.mode" somewhere but I'm not seeing it.  Also if globalSettings is ever assigned to be a new object does mode continue pointing to the old variable or does it get updated to the new one?

---

<sup>**[src/generic_ui/polymer/invite-user.html, line 75 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1916#-K-4FTdMrG4obQJGOSi7)** ([raw file](https://github.com/uproxy/uproxy/blob/3e3606ef0ada52d6d332354041007772afba6964/src/generic_ui/polymer/invite-user.html#L75)):</sup>
want to translate this before submititng, or add a TODO: translate?  (same for other new screens)

---

<sup>**[src/generic_ui/polymer/invite-user.ts, line 74 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1916#-K-4FeODz9SPy2L5V32I)** ([raw file](https://github.com/uproxy/uproxy/blob/3e3606ef0ada52d6d332354041007772afba6964/src/generic_ui/polymer/invite-user.ts#L74)):</sup>
Ideally this error would come from the GitHub social provider...  maybe make a TODO?

---

<sup>**[src/generic_ui/scripts/core_connector.ts, line 20 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1916#-K-4Ftsj_eptBo_BgMy7)** ([raw file](https://github.com/uproxy/uproxy/blob/3e3606ef0ada52d6d332354041007772afba6964/src/generic_ui/scripts/core_connector.ts#L20)):</sup>
want to remove this?

---

<sup>**[src/interfaces/social.ts, line 194 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1916#-K-4G9FsVLg6EkTa2GSY)** ([raw file](https://github.com/uproxy/uproxy/blob/3e3606ef0ada52d6d332354041007772afba6964/src/interfaces/social.ts#L194)):</sup>
These are kinda similar..  Can we change to something like "LOCAL_INVITED_BY_REMOTE" and "REMOTE_INVITED_BY_LOCAL"?  It's kinda long but better explains who does the inviting

---

<sup>**[src/interfaces/uproxy_core_api.ts, line 98 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1916#-K-4GPcSs4fMu1L2tIUt)** ([raw file](https://github.com/uproxy/uproxy/blob/3e3606ef0ada52d6d332354041007772afba6964/src/interfaces/uproxy_core_api.ts#L98)):</sup>
Changing these IDs is going to break this functionality for users who have different versions of the app, can we keep the old IDs for the ones which aren't changing?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1916)
<!-- Sent from Reviewable.io -->

lucyhe commented:
1) Good catch! It was actually a my other github account and I just hacked the username; didn't mean for the confusion
2) grunt test is passing for me now, will see what Travis says..

---

Review status: 0 of 29 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

<sup>**[package.json, line 34 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1916#-K-42SU0vo8F7mwYPxBw)** ([raw file](https://github.com/uproxy/uproxy/blob/3e3606ef0ada52d6d332354041007772afba6964/package.json#L34)):</sup>
Oh! Related to the TODO mentioned at the top: still need to publish the freedom-social-github package

---

<sup>**[src/generic_core/remote-user.ts, line 479 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1916#-K-44de2o02LMPo_Uklm)** ([raw file](https://github.com/uproxy/uproxy/blob/3e3606ef0ada52d6d332354041007772afba6964/src/generic_core/remote-user.ts#L479)):</sup>
Done.

---

<sup>**[src/generic_ui/polymer/accept-user-invite.ts, line 15 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1916#-K-47daXPVjVEM91eVJo)** ([raw file](https://github.com/uproxy/uproxy/blob/3e3606ef0ada52d6d332354041007772afba6964/src/generic_ui/polymer/accept-user-invite.ts#L15)):</sup>
Ah, you're right that I just did this to match the GitHub format. I changed this to just passing the networkName, since the current userId is determined in the core.

---

<sup>**[src/generic_ui/polymer/accept-user-invite.ts, line 22 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1916#-K-47bC6DqXK17zneMBa)** ([raw file](https://github.com/uproxy/uproxy/blob/3e3606ef0ada52d6d332354041007772afba6964/src/generic_ui/polymer/accept-user-invite.ts#L22)):</sup>
The nested one is the current user, the outer is your friend. But I no longer use the UserPath now.

---

<sup>**[src/generic_ui/polymer/contact.ts, line 118 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1916#-K-48bmOjpU0xR4f1eKZ)** ([raw file](https://github.com/uproxy/uproxy/blob/3e3606ef0ada52d6d332354041007772afba6964/src/generic_ui/polymer/contact.ts#L118)):</sup>
I changed it to pass mode from uproxy-roster-group -- now each contact knows whether it is a get or share contact. If globalSettings is changed, then all the roster groups would change as well, creating new contacts in the "template repeat" for contacts

---

<sup>**[src/generic_ui/scripts/core_connector.ts, line 20 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1916#-K-4Ftsj_eptBo_BgMy7)** ([raw file](https://github.com/uproxy/uproxy/blob/3e3606ef0ada52d6d332354041007772afba6964/src/generic_ui/scripts/core_connector.ts#L20)):</sup>
Done.

---

<sup>**[src/interfaces/social.ts, line 194 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1916#-K-4G9FsVLg6EkTa2GSY)** ([raw file](https://github.com/uproxy/uproxy/blob/3e3606ef0ada52d6d332354041007772afba6964/src/interfaces/social.ts#L194)):</sup>
Done.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1916)
<!-- Sent from Reviewable.io -->

lucyhe commented:
Review status: 0 of 30 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

<sup>**[src/generic_ui/polymer/invite-user.ts, line 0 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1916#-K-4FeODz9SPy2L5V32I)** ([raw file](https://github.com/uproxy/uproxy/blob/3e3606ef0ada52d6d332354041007772afba6964/src/generic_ui/polymer/invite-user.ts#L0)):</sup>
Done.

---

<sup>**[src/interfaces/uproxy_core_api.ts, line 98 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1916#-K-4GPcSs4fMu1L2tIUt)** ([raw file](https://github.com/uproxy/uproxy/blob/3e3606ef0ada52d6d332354041007772afba6964/src/interfaces/uproxy_core_api.ts#L98)):</sup>
Marked ADD_USER as deprecated, and added the new Commands at the end.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1916)
<!-- Sent from Reviewable.io -->

lucyhe commented:
Review status: 0 of 30 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

<sup>**[src/generic_ui/polymer/invite-user.html, line 0 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1916#-K-4FTdMrG4obQJGOSi7)** ([raw file](https://github.com/uproxy/uproxy/blob/3e3606ef0ada52d6d332354041007772afba6964/src/generic_ui/polymer/invite-user.html#L0)):</sup>
Done.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1916)
<!-- Sent from Reviewable.io -->

lucyhe commented:
Just linked the social provider via npm; ptal, thank you!

dborkan commented:
:+1: 
<!-- Reviewable comment -K-K6HVdZ4dk2gv9HhoK:-510466975 -->

---

Review status: 0 of 30 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

<sup>**[src/generic_core/remote-user.ts, line 479 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1916#-K-44de2o02LMPo_Uklm)** ([raw file](https://github.com/uproxy/uproxy/blob/3e3606ef0ada52d6d332354041007772afba6964/src/generic_core/remote-user.ts#L479)):</sup>
Do you agree there seems like a race condition here or am I nuts?
<!-- Reviewable comment -K-JouZlywAU8UtzJ_a9:-115908689 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1916)
<!-- Sent from Reviewable.io -->

lucyhe commented:
Review status: 0 of 30 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

<sup>**[src/generic_core/remote-user.ts, line 479 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1916#-K-44de2o02LMPo_Uklm)** ([raw file](https://github.com/uproxy/uproxy/blob/3e3606ef0ada52d6d332354041007772afba6964/src/generic_core/remote-user.ts#L479)):</sup>
I think you're exactly right. We might call update() before storage is done loading.
Another fix (in social.ts) seems to be 
var user = this.getOrAddUser_(userId);
user.onceLoaded(() => {
  user.update(profile);
});
<!-- Reviewable comment -K-PWmtiIyL3yXMVvGQE:-115219088 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1916)
<!-- Sent from Reviewable.io -->

dborkan commented:
Review status: 0 of 30 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

<sup>**[src/generic_core/remote-user.ts, line 479 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1916#-K-44de2o02LMPo_Uklm)** ([raw file](https://github.com/uproxy/uproxy/blob/3e3606ef0ada52d6d332354041007772afba6964/src/generic_core/remote-user.ts#L479)):</sup>
Cool, I'll enter a new issue for this since it wasn't added by this pull request and we haven't seen it breaking anything yet.
<!-- Reviewable comment -K-PZ2iNAhH20pXC82DA:586579350 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1916)
<!-- Sent from Reviewable.io -->
